organisations/tests/test_models.py

```diff
@@ -210,6 +215,85 @@ def test_too_old_personal_data(mock_get_event_data):
     assert all(person in results for person in expected_people)
 
 
+@pytest.mark.django_db
+def test_old_personal_data_deletion_with_valid_enrolments(mock_get_event_data):
+    too_old = timezone.now() - relativedelta(months=24, days=1)
+    still_valid = timezone.now() - relativedelta(days=1)
+
+    (
+        enrolment_person,
+        study_group_person,
+        enrolment_and_study_group_person,
+        no_enrolment_or_study_group_person,
+        event_queue_enrolment_person,
+        event_queue_study_group_person,
+    ) = PersonFactory.create_batch(6, user=None)
+
+    # person with a too old enrolment
+    EnrolmentFactory(person=enrolment_person, occurrence__end_time=too_old)
+
+    # person with a too old study group
+    EnrolmentFactory(
+        study_group__person=study_group_person, occurrence__end_time=too_old
+    )
+
+    # person with too old enrolment and study group
+    EnrolmentFactory(
+        person=enrolment_and_study_group_person,
+        study_group__person=enrolment_and_study_group_person,
+        occurrence__end_time=too_old,
+    )
+
+    too_old_p_event = PalvelutarjotinEventFactory()
+    OccurrenceFactory(end_time=too_old, p_event=too_old_p_event)
+
+    EventQueueEnrolmentFactory(
+        person=event_queue_enrolment_person, p_event=too_old_p_event
+    )
+    EventQueueEnrolmentFactory(
+        study_group__person=event_queue_study_group_person,
+        p_event=too_old_p_event,
+    )
+
+    valid_p_event = PalvelutarjotinEventFactory()
+    OccurrenceFactory(end_time=still_valid, p_event=valid_p_event)
+    # valid event queue enrolment for the event queue enrolment person
+    EventQueueEnrolmentFactory(
+        person=event_queue_enrolment_person, p_event=valid_p_event
+    )
+
+    # valid enrolment for the enrolment person
+    EnrolmentFactory(person=enrolment_person, occurrence__end_time=still_valid)
+    # valid enrolment for the study group person
+    EnrolmentFactory(
+        study_group__person=study_group_person, occurrence__end_time=still_valid
+    )
+    # valid enrolment for the event queue enrolment person
+    EventQueueEnrolmentFactory(
+        person=event_queue_enrolment_person, p_event=too_old_p_event
+    )
+    # valid enrolment for the event queue study group person
+    EventQueueEnrolmentFactory(
+        study_group__person=event_queue_study_group_person, p_event=too_old_p_event
+    )
+    # person with a user
+    PersonFactory(user=UserFactory())
+
+    expected_people = [
+        enrolment_and_study_group_person,
+        no_enrolment_or_study_group_person,
+        event_queue_study_group_person,
+    ]
+
+    # only persons without any valid enrolments should be returned
+    # enrolment_person and study_group_person have valid enrolments
+    # event_queue_enrolment_person has a valid event queue enrolment
+    # event_queue_study_group_person has a valid event queue enrolment
```

Comment:
Comment says "event_queue_study_group_person" has a valid event queue enrolment but they are in the list of "persons without any valid enrolments"?

Reply:
OK, looks like for this to hold the `retention_period_exceeded` should also consider `studygroup__queued_enrolments__p_event__occurrences__end_time__gt=earliest_valid_timestamp`